File: functions/src/didit-webhook.ts
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable camelcase */
/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/no-explicit-any */
const functions = require("firebase-functions");
const admin = require("firebase-admin");
const crypto = require("crypto");

// Inicializa o Firebase Admin (apenas uma vez)
if (!admin.apps.length) {
  admin.initializeApp();
}

const db = admin.firestore();

/**
 * Webhook do Didit para receber notificações de verificação
 *
 * Endpoint: https://us-central1-partiu-479902.cloudfunctions.net/diditWebhook
 *
 * Funcionalidades:
 * - Valida assinatura HMAC do webhook
 * - Verifica timestamp (máximo 5 minutos)
 * - Processa eventos de status e dados
 * - Atualiza sessão no Firestore
 * - Salva verificação aprovada automaticamente
 */
exports.diditWebhook = functions.https.onRequest(async (req: any, res: any) => {
  // Configurar CORS se necessário
  res.set("Access-Control-Allow-Origin", "*");

  if (req.method === "OPTIONS") {
    res.set("Access-Control-Allow-Methods", "POST");
    res.set("Access-Control-Allow-Headers", "Content-Type, X-Signature, X-Timestamp");
    return res.status(204).send("");
  }

  // Apenas aceita POST
  if (req.method !== "POST") {
    return res.status(405).json({error: "Method not allowed"});
  }

  try {
    // 1. Buscar Webhook Secret do Firestore
    const configDoc = await db.collection("AppInfo").doc("didio").get();

    if (!configDoc.exists) {
      console.error("Configuração do Didit não encontrada em AppInfo/didio");
      return res.status(500).json({error: "Configuration not found"});
    }

    const config = configDoc.data();
    const WEBHOOK_SECRET = config.webhook_secret;

    if (!WEBHOOK_SECRET) {
      console.error("Webhook secret não configurado em AppInfo/didio");
      return res.status(500).json({error: "Webhook secret not configured"});
    }

    // 2. Extrair headers de segurança
    const signature = req.get("X-Signature");
    const timestamp = req.get("X-Timestamp");

    if (!signature || !timestamp) {
      console.error("Headers de segurança ausentes");
      return res.status(401).json({error: "Missing security headers"});
    }

    // 3. Validar timestamp (máximo 5 minutos)
    const currentTime = Math.floor(Date.now() / 1000);
    const incomingTime = parseInt(timestamp, 10);

    if (Math.abs(currentTime - incomingTime) > 300) {
      console.error("Timestamp expirado:", {currentTime, incomingTime});
      return res.status(401).json({error: "Request timestamp is stale"});
    }

    // 4. Obter o body raw (já está em req.rawBody no Firebase Functions)
    // IMPORTANTE: Usar req.rawBody para garantir que a assinatura corresponda exatamente ao que foi enviado
    const rawBody = req.rawBody || JSON.stringify(req.body);

    // 5. Validar assinatura HMAC
    const hmac = crypto.createHmac("sha256", WEBHOOK_SECRET);
    const expectedSignature = hmac.update(rawBody).digest("hex");

    const expectedBuffer = Buffer.from(expectedSignature, "utf8");
    const providedBuffer = Buffer.from(signature, "utf8");

    if (
      expectedBuffer.length !== providedBuffer.length ||
      !crypto.timingSafeEqual(expectedBuffer, providedBuffer)
    ) {
      console.error("Assinatura inválida:", {
        expected: expectedSignature,
        provided: signature,
      });
      return res.status(401).json({error: "Invalid signature"});
    }

    // 6. Processar o webhook
    const webhookData = req.body;
    const {
      session_id,
      status,
      webhook_type,
      vendor_data,
      workflow_id,
      metadata,
      decision,
      created_at,
      timestamp: webhookTimestamp,
    } = webhookData;

    console.log("Webhook recebido:", {
      session_id,
      status,
      webhook_type,
      vendor_data,
    });

    // Verificar duplicidade (Idempotência)
    if (webhookTimestamp) {
      const existingWebhook = await db.collection("DiditWebhooks")
        .where("session_id", "==", session_id)
        .where("timestamp", "==", webhookTimestamp)
        .limit(1)
        .get();

      if (!existingWebhook.empty) {
        console.log("Webhook duplicado ignorado:", session_id, webhookTimestamp);
        return res.status(200).json({
          message: "Webhook already processed (duplicate)",
          session_id: session_id,
        });
      }
    }

    // 7. Salvar webhook no histórico (apenas se for status relevante)
    // Ignora status intermediários para evitar poluição do banco
    const relevantStatuses = ["Approved", "Rejected", "Failed", "In Review"];
    let docRef: any = null;

    if (relevantStatuses.includes(status)) {
      const webhookRecord: any = {
        session_id,
        status,
        webhook_type,
        vendor_data,
        received_at: admin.firestore.FieldValue.serverTimestamp(),
        processed: false,
      };

      // Adiciona apenas campos que não são undefined
      if (workflow_id !== undefined) webhookRecord.workflow_id = workflow_id;
      if (metadata !== undefined) webhookRecord.metadata = metadata;
      if (decision !== undefined) webhookRecord.decision = decision;
      if (created_at !== undefined) webhookRecord.created_at = created_at;
      if (webhookTimestamp !== undefined) webhookRecord.timestamp = webhookTimestamp;

      docRef = await db.collection("DiditWebhooks").add(webhookRecord);
    } else {
      console.log(`Webhook com status '${status}' não salvo em DiditWebhooks (ignorado)`);
    }

    // 8. Atualizar sessão no Firestore
    const sessionRef = db.collection("DiditSessions").doc(session_id);
    const sessionDoc = await sessionRef.get();

    if (!sessionDoc.exists) {
      console.warn("Sessão não encontrada:", session_id);
      // Não retorna erro, webhook é válido mesmo se sessão não existir
    } else {
      const updateData: any = {
        status: status,
        lastWebhookType: webhook_type,
        lastWebhookAt: admin.firestore.FieldValue.serverTimestamp(),
      };

      // Apenas marca como completedAt se for status final
      if (status === "Approved" || status === "Rejected" || status === "Failed") {
        updateData.completedAt = admin.firestore.FieldValue.serverTimestamp();
      }

      // Adiciona decision se existir
      if (decision !== undefined) {
        updateData.result = decision;
      }

      await sessionRef.update(updateData);

      console.log("Sessão atualizada:", session_id, "status:", status);
    }

    // 9. Se aprovado, salvar verificação
    if (status === "Approved" && decision && decision.id_verification) {
      const idVerification = decision.id_verification;
      const userId = vendor_data; // vendor_data é o userId

      if (userId && idVerification.status === "Approved") {
        try {
          // Salvar em FaceVerifications
          await db.collection("FaceVerifications").doc(userId).set({
            userId: userId,
            facialId: session_id,
            verifiedAt: admin.firestore.FieldValue.serverTimestamp(),
            status: "verified",
            gender: idVerification.gender || null,
            age: idVerification.age || null,
            details: {
              verification_type: "didit",
              verification_date: new Date().toISOString(),
              document_type: idVerification.document_type,
              document_number: idVerification.document_number,
              full_name: idVerification.full_name,
              first_name: idVerification.first_name,
              last_name: idVerification.last_name,
              date_of_birth: idVerification.date_of_birth,
              nationality: idVerification.nationality,
              issuing_state: idVerification.issuing_state_name,
              portrait_image: idVerification.portrait_image,
              session_id: session_id,
              session_url: decision.session_url,
            },
          });

          // Atualizar usuário
          await db.collection("Users").doc(userId).set({
            user_is_verified: true,
            verified_at: admin.firestore.FieldValue.serverTimestamp(),
            facial_id: session_id,
            verification_type: "didit",
          }, {merge: true});

          console.log("Verificação salva automaticamente para:", userId);
        } catch (error) {
          console.error("Erro ao salvar verificação:", error);
          // Não retorna erro, webhook foi processado com sucesso
        }
      }
    }

    // 10. Marcar webhook como processado (se foi salvo)
    if (docRef) {
      await docRef.update({processed: true});
    }

    // 11. Retornar sucesso
    return res.status(200).json({
      message: "Webhook processed successfully",
      session_id: session_id,
      status: status,
    });
  } catch (error) {
    console.error("Erro ao processar webhook:", error);
    return res.status(500).json({
      error: "Internal server error",
      message: (error as any).message,
    });
  }
});

File: lib/core/models/didit_webhook.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// Modelo para decisão de verificação do Didit
class DiditDecision {
  final String sessionId;
  final int sessionNumber;
  final String sessionUrl;
  final String status;
  final String workflowId;
  final List<String> features;
  final String? vendorData;
  final Map<String, dynamic>? metadata;
  final DiditIdVerification? idVerification;
  final List<DiditReview>? reviews;
  final DateTime createdAt;

  DiditDecision({
    required this.sessionId,
    required this.sessionNumber,
    required this.sessionUrl,
    required this.status,
    required this.workflowId,
    required this.features,
    this.vendorData,
    this.metadata,
    this.idVerification,
    this.reviews,
    required this.createdAt,
  });

  factory DiditDecision.fromJson(Map<String, dynamic> json) {
    return DiditDecision(
      sessionId: json['session_id'] as String,
      sessionNumber: json['session_number'] as int,
      sessionUrl: json['session_url'] as String,
      status: json['status'] as String,
      workflowId: json['workflow_id'] as String,
      features: (json['features'] as List<dynamic>).cast<String>(),
      vendorData: json['vendor_data'] as String?,
      metadata: json['metadata'] as Map<String, dynamic>?,
      idVerification: json['id_verification'] != null
          ? DiditIdVerification.fromJson(
              json['id_verification'] as Map<String, dynamic>)
          : null,
      reviews: json['reviews'] != null
          ? (json['reviews'] as List<dynamic>)
              .map((r) => DiditReview.fromJson(r as Map<String, dynamic>))
              .toList()
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'session_id': sessionId,
      'session_number': sessionNumber,
      'session_url': sessionUrl,
      'status': status,
      'workflow_id': workflowId,
      'features': features,
      if (vendorData != null) 'vendor_data': vendorData,
      if (metadata != null) 'metadata': metadata,
      if (idVerification != null) 'id_verification': idVerification!.toJson(),
      if (reviews != null) 'reviews': reviews!.map((r) => r.toJson()).toList(),
      'created_at': createdAt.toIso8601String(),
    };
  }
}

/// Modelo para verificação de ID do Didit
class DiditIdVerification {
  final String status;
  final String? documentType;
  final String? documentNumber;
  final String? personalNumber;
  final String? portraitImage;
  final String? frontImage;
  final String? backImage;
  final String? dateOfBirth;
  final int? age;
  final String? expirationDate;
  final String? dateOfIssue;
  final String? issuingState;
  final String? issuingStateName;
  final String? firstName;
  final String? lastName;
  final String? fullName;
  final String? gender;
  final String? address;
  final String? nationality;
  final List<Map<String, dynamic>>? warnings;

  DiditIdVerification({
    required this.status,
    this.documentType,
    this.documentNumber,
    this.personalNumber,
    this.portraitImage,
    this.frontImage,
    this.backImage,
    this.dateOfBirth,
    this.age,
    this.expirationDate,
    this.dateOfIssue,
    this.issuingState,
    this.issuingStateName,
    this.firstName,
    this.lastName,
    this.fullName,
    this.gender,
    this.address,
    this.nationality,
    this.warnings,
  });

  factory DiditIdVerification.fromJson(Map<String, dynamic> json) {
    return DiditIdVerification(
      status: json['status'] as String,
      documentType: json['document_type'] as String?,
      documentNumber: json['document_number'] as String?,
      personalNumber: json['personal_number'] as String?,
      portraitImage: json['portrait_image'] as String?,
      frontImage: json['front_image'] as String?,
      backImage: json['back_image'] as String?,
      dateOfBirth: json['date_of_birth'] as String?,
      age: json['age'] as int?,
      expirationDate: json['expiration_date'] as String?,
      dateOfIssue: json['date_of_issue'] as String?,
      issuingState: json['issuing_state'] as String?,
      issuingStateName: json['issuing_state_name'] as String?,
      firstName: json['first_name'] as String?,
      lastName: json['last_name'] as String?,
      fullName: json['full_name'] as String?,
      gender: json['gender'] as String?,
      address: json['address'] as String?,
      nationality: json['nationality'] as String?,
      warnings: json['warnings'] != null
          ? (json['warnings'] as List<dynamic>).cast<Map<String, dynamic>>()
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'status': status,
      if (documentType != null) 'document_type': documentType,
      if (documentNumber != null) 'document_number': documentNumber,
      if (personalNumber != null) 'personal_number': personalNumber,
      if (portraitImage != null) 'portrait_image': portraitImage,
      if (frontImage != null) 'front_image': frontImage,
      if (backImage != null) 'back_image': backImage,
      if (dateOfBirth != null) 'date_of_birth': dateOfBirth,
      if (age != null) 'age': age,
      if (expirationDate != null) 'expiration_date': expirationDate,
      if (dateOfIssue != null) 'date_of_issue': dateOfIssue,
      if (issuingState != null) 'issuing_state': issuingState,
      if (issuingStateName != null) 'issuing_state_name': issuingStateName,
      if (firstName != null) 'first_name': firstName,
      if (lastName != null) 'last_name': lastName,
      if (fullName != null) 'full_name': fullName,
      if (gender != null) 'gender': gender,
      if (address != null) 'address': address,
      if (nationality != null) 'nationality': nationality,
      if (warnings != null) 'warnings': warnings,
    };
  }

  /// Verifica se a verificação foi aprovada
  bool get isApproved => status == 'Approved';
}

/// Modelo para review de verificação
class DiditReview {
  final String user;
  final String newStatus;
  final String? comment;
  final DateTime createdAt;

  DiditReview({
    required this.user,
    required this.newStatus,
    this.comment,
    required this.createdAt,
  });

  factory DiditReview.fromJson(Map<String, dynamic> json) {
    return DiditReview(
      user: json['user'] as String,
      newStatus: json['new_status'] as String,
      comment: json['comment'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user': user,
      'new_status': newStatus,
      if (comment != null) 'comment': comment,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

/// Modelo para webhook do Didit
class DiditWebhook {
  final String sessionId;
  final String status;
  final String webhookType;
  final int createdAt;
  final int timestamp;
  final String? workflowId;
  final String? vendorData;
  final Map<String, dynamic>? metadata;
  final DiditDecision? decision;

  DiditWebhook({
    required this.sessionId,
    required this.status,
    required this.webhookType,
    required this.createdAt,
    required this.timestamp,
    this.workflowId,
    this.vendorData,
    this.metadata,
    this.decision,
  });

  factory DiditWebhook.fromJson(Map<String, dynamic> json) {
    return DiditWebhook(
      sessionId: json['session_id'] as String,
      status: json['status'] as String,
      webhookType: json['webhook_type'] as String,
      createdAt: json['created_at'] as int,
      timestamp: json['timestamp'] as int,
      workflowId: json['workflow_id'] as String?,
      vendorData: json['vendor_data'] as String?,
      metadata: json['metadata'] as Map<String, dynamic>?,
      decision: json['decision'] != null
          ? DiditDecision.fromJson(json['decision'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'session_id': sessionId,
      'status': status,
      'webhook_type': webhookType,
      'created_at': createdAt,
      'timestamp': timestamp,
      if (workflowId != null) 'workflow_id': workflowId,
      if (vendorData != null) 'vendor_data': vendorData,
      if (metadata != null) 'metadata': metadata,
      if (decision != null) 'decision': decision!.toJson(),
    };
  }

  /// Converte para Firestore
  Map<String, dynamic> toFirestore() {
    return {
      ...toJson(),
      'received_at': FieldValue.serverTimestamp(),
    };
  }

  /// Verifica se é um webhook de status atualizado
  bool get isStatusUpdated => webhookType == 'status.updated';

  /// Verifica se é um webhook de dados atualizados
  bool get isDataUpdated => webhookType == 'data.updated';

  /// Verifica se a verificação foi aprovada
  bool get isApproved => status == 'Approved';

  /// Verifica se a verificação foi recusada
  bool get isDeclined => status == 'Declined';

  /// Verifica se a verificação está em revisão
  bool get isInReview => status == 'In Review';

  /// Verifica se possui decisão
  bool get hasDecision => decision != null;

  /// Verifica se possui verificação de ID aprovada
  bool get hasApprovedIdVerification =>
      decision?.idVerification?.isApproved ?? false;
}
